File: 백준/Silver/16194. 카드 구매하기 2/카드 구매하기 2.py
def main(arr):
    dp = [float('inf')] * (len(arr) + 2)

    for i in range(1, len(arr) + 1):
        dp[i] = min(arr[i - 1], dp[i])
        for j in range(i + 1, len(arr) + 1):
            dp[j] = min(dp[j - i] + dp[i], dp[j])

    return dp[len(arr)]


n = int(input())
arr = [int(i) for i in input().split()]
print(main(arr))

File: 백준/Silver/16194. 카드 구매하기 2/카드 구매하기 2.js
const main = (arr) => {
    const dp = [0, ...arr]
    
    for(let i = 1; i <= arr.length; i++){
        for (let j = i + 1; j <= arr.length; j ++){
            dp[j] = Math.min(dp[j - i] + dp[i], dp[j])
        }
    }
    
    return dp[arr.length];
}

const fs = require('fs');
let [n, arr] = fs.readFileSync('/dev/stdin').toString().trim().split("\n");
arr = arr.split(" ").map(Number)

console.log(main(arr))

